src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -573,12 +573,51 @@ private Task InvokeAllExceptionFiltersAsync()
 
                         _logger.ActionMethodExecuting(_actionExecutingContext, arguments);
 
-                        var actionReturnValue = await ControllerActionExecutor.ExecuteAsync(
-                            _executor,
-                            _controller,
-                            arguments);
+                        var returnType = _executor.MethodReturnType;
 
-                        result = CreateActionResult( actionMethodInfo.ReturnType, actionReturnValue);
+                        if (returnType == typeof(void))
+                        {
+                            _executor.Execute(_controller, arguments);
+                            result = new EmptyResult();
+                        }
+                        else if (returnType == typeof(IActionResult))
```

Comment:
What if you check `IsMethodAsync` and `IsTypeAssignableFromIActionResult` right here, and throw the exception for null action results right here.

Same comment for  `_executor.TaskGenericType == typeof(IActionResult)` - it will read better and you don't have to add complexity to `CreateActionResult`